### ISSUE TITLE
Zenith: Confusion over what "optional" means

### ISSUE THREAD

**Issue by helderco** [dagger/dagger#6749]
## Summary

While documenting optional and default values in Python functions I came upon some inconsistencies in implementations due to different assumptions. The term “Optional” that’s used in modules is misleading.

## Terminology

In Python, an argument that is optional is one with a default value because **the true meaning of “optional” is that it can be omitted**.

```python
def foo(bar: str = ""): ...
```

In GraphQL it’s mostly the same, with the exception that a **nullable field without a default can also be omitted**, in which case it automatically defaults to `null`:

```graphql
foo: String  # null if omitted
```

The same doesn’t happen in Python:

```python
def foo(bar: str | None): ...  # this is a required argument!
```

But in the Dagger API, we use “optional” to mean nullable. I think people usually find the nullable/non-null terminology a bit strange and tend to prefer thinking about the “!” at the end of a type as meaning “required” (me included):

```graphql
foo: String!  # does not mean required! it means null is not allowed
```

So some may find it surprising that you can actually have a type that is non-null and has a default:

```graphql
foo: String! = "bar"  # this is actually optional due to the default!
```

## Core schema generation

Our core API has only nullable fields when a default is used:

```graphql
foo: String = "bar"
```

There’s a use case for that, especially in terms of being able to express the difference between “omitted” and “unset”:
- **Not provided**: Use the default “bar” string.
- **Provided as `null`**: Explicitly remove or clear foo.
- **Provided as a non-null string**: Update foo to the provided value.

But I think, at least in most cases, they should be *non-null* instead. 

In Go, where the schema is expressed in, to be nullable, the type should be a pointer, that allows for *nil*. Descriptions are optional, but being a non-null empty string is enough. There is no advantage in also allowing a `null` to clear this value.

For example, this is how core generates the following argument:

```go
Description string `default:""`  // description: String = ""
```

But I think it should be this:

```go
Description string `default:""`  // description: String! = ""
Description *string `default:""`  // description: String = ""
```

\cc @vito 

## SDK inconsistency

Let’s see how the SDKs can express the following GraphQL schema:

```graphql
foo(
    a: String!,
    b: String,
    c: String! = "foo",
    d: String = null,
    e: String = "bar"
): String!
```

### Go

Go modules can currently express all of them, it’s just the terminology of having “optional” independent from “default” that can be confusing:

```go
type Foo struct{ }

func (m *Foo) Bar(
	a string,
	// +optional
	b *string,
	// +default="foo"
	c string,
	// +optional
	// +default=null
	d *string,
	// +optional
	// +default="bar"
	e *string,
) string {
	return fmt.Sprintf("%+v, %+v, %+v, %+v, %+v", a, b, c, d, e)
}
```

### Python

Python doesn’t allow `str | None` without a default (returns a validation error since v0.9.9), and simply equated having a default to being optional, but I’ve fixed that (locally) to match GraphQL:

```python
@object_type
class Foo:
    @function
    def foo(
        self,
        a: str,
        b: str | None,
        c: str = "hello",
        d: str | None = None,
        e: str | None = "hello",
    ) -> str:
        return ", ".join(repr(x) for x in (a, b, c, d, e))
```

To match GraphQL’s behavior of defaulting to `null` on nullable types, I’ve decided to have the SDK automatically set `None` for an omitted nullable argument. However, when called from Python code, `str | None` without a default is still required, so you need to explicitly pass `None` to it.

### TypeScript

It’s not possible in TypeScript today to express a non-null argument with a default:

```typescript
@object()
class Foo {
  @func()
  foo(
    a: string,
    b?: string,  // or b: string | null
    c: string = "foo",  // ❌ c: String = "foo"
    d: string | null = null,
    e: string | null = "bar",
  ): string {
    return `${a}, ${b}, ${c}, ${d}, ${e}`
  }
}
```

I suppose it’s for the same reason as Python where having a default equates to being optional, but we should change it as well. 

\cc @TomChv

## Query values

It seems I’m not able to send a `null` value to nullable types. The Engine simply omits the argument if the value is `null`. So if it’s a nullable type with a default, `null` can’t be used to differentiate between “omitted” and “unset”, for example.

However, I can’t think of a use case for this, just exploring expressiveness in the API and SDKs.

\cc @sipsma

**Comment by TomChv** [comment]
> I suppose it’s for the same reason as Python where having a default equates to being optional, but we should change it as well.

Yes exactly, which kind of make sense in the context of our language but I agree that it's not aligned with GraphQL.

**Comment by helderco** [comment]
Note: It's a bit late to change the meaning of Optional, so I'm just trying to make things consistent (between API, GraphQL and SDKs).

**Comment by helderco** [comment]
Update: I haven't been able to circle back to this yet but there's still some inconsistencies in core and I think we should change the name "optional" in the SDK internal APIs, but I need to invest some time in thinking more about the way we use nulls and whether there's value in being able to express non-nulls with defaults, for example, or ditch it in favor of more intuitive DX in the SDKs.

**Comment by helderco** [comment]
## Go SDK

Yeah, those changes to Go sgtm. One addendum:

> Behavior on schema generation: using `+default` without `+optional` generates non-nullable argument with a default value.

Unless the default is `null`, in which case it generates a nullable argument.

Also, how would you define `e: String = "bar"`? According to this definition you’d need both `+default` and `+optional` but that brings back `+optional` to mean *nullable* again and it’s not likely to be clear to users.

That takes me to the next question on supporting *null* values. But before that, just to clarify, to produce the example schema above:

```graphql
foo(
    a: String!,
    b: String,
    c: String! = "foo",
    d: String = null,
    e: String = "bar"
): String!
```

Would become:

```diff
 type Foo struct{ }

 func (m *Foo) Bar(
        a string,
        // +optional
-       b *string,
+       b string,
        // +default="foo"
        c string,
-       // +optional
        // +default=null
-       d *string,
+       d string,
        // +optional
        // +default="bar"
-       e *string,
+       e string,
 ) string {
        return fmt.Sprintf("%+v, %+v, %+v, %+v, %+v", a, b, c, d, e)
 }
```

We could also not add the `= null` since it’s redundant (from any SDK).
## Do we want to support `null` values?

See sections “Core schema generation” and “Query values” in the description for more context.

Last time I worked on this I was concerned with full expressiveness from the SDKs into the API schema. But after making sure an SDK correctly produces a nullable type, I was not able to use a `null` value anywhere. That’s because **the API server currently equates a null value to an omitted value**. No actual *null* value is passed around.

So the question here is **do we make sure we match GraphQL and allow null values?** Or do we not have a use for it in Dagger and might as well not worry about full expressiveness in the SDKs and simplify them to match the same meaning in core?

The answer to this question can potentially change the core API schema and SDKs, but it doesn’t necessarily mean in a way that users will notice. However, it’s necessary to move forward in bringing consistency everywhere.

**Comment by TomChv** [comment]
@helderco What are the current progress with this issue? 
This proposal leads to severals changes in our SDKs I think so it would be better if we do it inside a single PR.

**Comment by helderco** [comment]
It's ok to be different PRs:
- Go changes around `+default`
- Core API handling of null values
- Internal rename of `withOptional` in type defs

Haven't had time to get back to this yet. I still have questions around our handling of null values in the API, see my last comment. Requesting feedback on that.

**Comment by helderco** [comment]
Update: had a discussion with the team and it might be a bug that the engine doesn't accept the null values, so we'll investigate where that comes from and go from there. Seems like passing `null` is something we should allow.

**Comment by helderco** [comment]
One fix for Go's codegen here:
- https://github.com/dagger/dagger/pull/8106
